File: Problems/0984_String_Without_AAA_or_BBB/Project_CS/String_Without_AAA_or_BBB.cs
using System;

public class Solution
{
    public string StrWithout3a3b(int A, int B)
    {
        string resultStr = "";

        while (A > B && B > 0)
        {
            resultStr += "aab";
            A -= 2;
            B--;
        }

        while (B > A && A > 0)
        {
            resultStr += "bba";
            B -= 2;
            A--;
        }

        if (A == 0 || B == 0)
            if (B == 0)
                resultStr += new string('a', A);
            else
                resultStr += new string('b', B);
        else if (resultStr.Length == 0 || resultStr[resultStr.Length - 1] == 'a')
        {
            // resultStr += new string("ba", A);
            for (int i = 0; i < A; ++i)
                resultStr += "ba";
        }
        else
        {
            // resultStr += new string("ab", A);
            for (int i = 0; i < B; ++i)
                resultStr += "ab";
        }

        return resultStr;
    }

    public void Main(string args)
    {
        string[] flds = args.Replace("[","").Replace("]","").Trim().Split(',');
        int A = int.Parse(flds[0]);
        int B = int.Parse(flds[1]);

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        string result = StrWithout3a3b(A, B);
        Console.WriteLine("result = " + result);

        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}

File: Problems/0900_0999/0984_String_Without_AAA_or_BBB/Project_C/String_Without_AAA_or_BBB.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>
#include <time.h>

#include "../../mylib_C/mylib.h"

/* Function prototype declaration */
char* strWithout3a3b(int A, int B);
int loop_main(char* arg);

#define RESULT_MAX_SIZE    201

char* strWithout3a3b(int A, int B)
{
    char* resultStr = (char *)malloc(sizeof(char)*RESULT_MAX_SIZE);
    resultStr[0] = '\0';

    if (A + B > RESULT_MAX_SIZE)
        err_exit("strWithout3a3b() resultStr size over.\n");

    while (A > B && B > 0) {
        strcat(resultStr, "aab");
        A -= 2;
        B--;
    }

    while (B > A && A > 0) {
        strcat(resultStr, "bba");
        B -= 2;
        A--;
    }

    if (A == 0 || B == 0) {
        if (B == 0) {        
            for (int i = 0; i < A; ++i)
                strcat(resultStr, "a");
        } else {
            for (int i = 0; i < B; ++i)
                strcat(resultStr, "b");
        }
    } else if (strlen(resultStr) == 0 || resultStr[strlen(resultStr) - 1] == 'a') {
        for (int i = 0; i < A; ++i)
            strcat(resultStr, "ba");
    } else {
        for (int i = 0; i < B; ++i)
            strcat(resultStr, "ab");
    }

    if (strlen(resultStr) > RESULT_MAX_SIZE)
        err_exit("strWithout3a3b() resultStr size over.\n");

    return resultStr;    
}

int loop_main(char* arg)
{
    int argv_length = strlen(arg);
    char* flds[2];
    char* str_nums[256];
    int nums[256];

    replace(arg, "[", "");
    replace(arg, "]", "");
    replace(arg, "\n", "");
    int flds_length = split(arg, ",", flds, sizeof(flds)/sizeof(flds[0]));
    int A = atoi(flds[0]);
    int B = atoi(flds[1]);

    clock_t time_start = clock();
    char* results = strWithout3a3b(A, B);
    clock_t time_end = clock();

    // result print.
    printf("result = %s\n", results);
    printf("Execute time ... %.0f ms\n\n", 1000*(double)(time_end - time_start)/CLOCKS_PER_SEC);

    // int* results clear.
    free(results);

    // char* flds[] free().
    p_char_array_free(flds, flds_length);

    return 0;
}

int main(int argc, char* argv[])
{
#define fgets_MAX   65536
    FILE *fp;
    char line[fgets_MAX];

    if (argc < 2) {
        printf("Usage %s <testdatafile>\n", argv[0]);
        return -1;
    }

    // File Open
    fp = fopen(argv[1], "r");

    if (fp == NULL) {
        printf("%s Open Failed...\n", argv[1]);
        return -1;
    }

    while((fgets(line, fgets_MAX - 1, fp)) != NULL) {
        trim(line);
        if (*line == '\0')
            continue;
        printf("arg = %s\n", line);
        loop_main(line);
    }

    fclose(fp);
    return 0;
}
